Add --base-hosts-file flag

gavinkflam opened the pull request:
I have added a container level base-hosts-file configuration in #21013. This PR allows user to apply this configuration using a CLI flag.

<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

In case you're only changing docs, make sure to prefix the pull-request title with "[CI:DOCS]". That will prevent functional tests from running and save time and energy.

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
Add --base-hosts-file flag to container create, container run, and pod create
```

gavinkflam commented:
I have two assumptions right now. Does it make sense?

1. `podman run --pod name` doesn't support `--base-hosts-file` because containers inherit hosts file from the pod

Using `--base-hosts-file` and `--pod`/`--pod-id-file` should result in a conflict and an error message?

2. `--no-hosts` has a higher precedence because it tells Podman not to manage the hosts file

Using `--base-hosts-file` and `--no-hosts` should result in a conflict and an error message?

gavinkflam commented:
I have a few things that I'm not so sure.

1. `podman run --pod name` doesn't support `--base-hosts-file` because containers inherit hosts file from the pod

I can either try to support this scenario, or put an error message banning this use case. I.e. what we are currently doing with `--add-host` and `--no-hosts`.

2. `--no-hosts` invalidates any `--base-hosts-file` values because it tells Podman not to manage the hosts file at all

`--no-hosts` and `--base-hosts-file=image` have different effects. `--no-hosts` is preventing Podman from adding entries derived from the network and the container hostname at all. It makes no sense to use them together though. An error message is probably good enough?

3. --base-hosts-file=/path/to/hosts is not working with remote connection

I can probably look into how other flags implement this. E.g. `--authfile`

rhatdan commented on cmd/podman/common/netflags.go:
```diff
@@ -26,6 +26,13 @@ func DefineNetFlags(cmd *cobra.Command) {
 	)
 	_ = cmd.RegisterFlagCompletionFunc(addHostFlagName, completion.AutocompleteNone)
 
+	baseHostsFileFlagName := "base-hosts-file"
```
I would just call tis hosts-file Make it much easier to find for the users.

Luap99 replied:
I am fine with hosts-file as name.

rhatdan commented on cmd/podman/common/completion.go:
```diff
@@ -964,6 +964,13 @@ func AutocompleteUserNamespace(cmd *cobra.Command, args []string, toComplete str
+// AutocompleteBaseHostsFile - Autocomplete base hosts file options.
+// -> "image", "none", paths
+func AutocompleteBaseHostsFile(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	BaseHostsFileModes := []string{"image", "none"}
```
I don't see a reason for image it is equivalent to "".  "none" conflicts with --no-hosts, if you want to replace --no-hosts with this and make --no-hosts hidden, I would be fine with that.

rhatdan commented on cmd/podman/common/netflags.go:
```diff
@@ -26,6 +26,13 @@ func DefineNetFlags(cmd *cobra.Command) {
+	baseHostsFileFlagName := "base-hosts-file"
+	netFlags.String(
+		baseHostsFileFlagName, "",
+		`Path to a hosts file to copy the entries into the container, or one of the special values. ("image"|"none")`,
```
I don't like it copying in, it should replace the entries.  The values should be 

"" (Default), "none"==--no-hosts, /PATHTO/hosts file.

gavinkflam commented:
Hi @rhatdan, thank you for reviewing the changes.

My idea is to provide a container and pod level override of `base_hosts_file` in [`containers.conf`](https://github.com/containers/common/blob/main/docs/containers.conf.5.md#containers-table), which is currently defaulted to the host's `/etc/hosts`. I agree `--no-hosts` seems redundant with the introduction of this new flag. I can hide it if you think that is appropriate.

| Flag | Container hosts file |
| ------ | ------ |
| `--base-hosts-file=""` | Entries in `base_hosts_file` in `containers.conf` + Podman managed entries |
| `--base-hosts-file="/path/to/hosts"` | Entries in `/path/to/hosts` + Podman managed entries |
| `--base-hosts-file="image"` | Entries in the image's `/etc/hosts` + Podman managed entries |
| `--base-hosts-file="none"` | Use only Podman managed entries |
| `--no-hosts` | Use the image `/etc/hosts` without adding any entries |

The current `--no-hosts` flag is similar to `--base-hosts-file="image"` with a subtle difference that `--no-hosts` doesn't add Podman managed entries. They include container name, `add_hosts` in `containers.conf`, network host entries, pasta IP, etc.

https://github.com/containers/podman/blob/7dc7cbfd9b0440ddc86e210a2272fdaccd6376bb/libpod/container_internal_common.go#L2282-L2288

Do you think this fits into the wider Podman design and roadmap? I am open to changing anything or even starting from scratch.

rhatdan commented:
I am fine with this, but I do not think `base-` adds anything to my understanding.  If we can find a  way to eliminate --no-host at all and do

--hosts-file="null" 	Use only Podman managed entries
--hosts-file="none"      Use the image /etc/hosts without adding any entries (--no-hosts)

Bottom line for me is discoverability for these. Having them all named and documented in the man page together would help users figure out which is best for them.

Having a random --no-hosts someone else in the man page, does not help with this.

Luap99 commented on cmd/podman/common/completion.go:
```diff
@@ -964,6 +964,13 @@ func AutocompleteUserNamespace(cmd *cobra.Command, args []string, toComplete str
+// -> "image", "none", paths
+func AutocompleteBaseHostsFile(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	BaseHostsFileModes := []string{"image", "none"}
+	return BaseHostsFileModes, cobra.ShellCompDirectiveNoSpace
```
```suggestion
	return BaseHostsFileModes, cobra.ShellCompDirectiveDefault
```

Luap99 commented on docs/source/markdown/options/base-hosts-file.md:
```diff
@@ -0,0 +1,9 @@
+
+BaseHostsFile is the path to a hosts file, the entries from this file are added to the <<containers|pods>> _/etc/hosts_ file.
+As special value "image" is allowed which uses the _/etc/hosts_ file from within the image and "none" which uses no base file at all.
+If it is empty we should default to the base_hosts_file configuration in _containers.conf_.
```
```suggestion
If it is empty it defaults to the base_hosts_file configuration in _containers.conf_ and if this is not set is uses `/etc/hosts` on the host system.
```

Luap99 commented on test/e2e/run_test.go:
```diff
@@ -1483,6 +1483,22 @@ VOLUME %s`, ALPINE, volPath, volPath)
+		confFile := filepath.Join(podmanTest.TempDir, "containers.conf")
+		err = os.WriteFile(confFile, []byte("[containers]\nbase_hosts_file=\"none\"\n"), 0755)
+		Expect(err).ToNot(HaveOccurred())
+		os.Setenv("CONTAINERS_CONF", confFile)
```
use CONTAINERS_CONF_OVERRIDE

Luap99 reviewed (commented):
`--no-hosts` is a well known docker option so we should not remove it from the docs and this option does not replace this option at all.

We should make the connection between the two options clear int he man page, i.e. --no-hosts should link to --hosts-file and the other way around.

gavinkflam commented:
@rhatdan @Luap99 Sounds like we have two directions here.

1. Bring together `--no-hosts` and this new flag, and introduce a new mode to use only Podman managed entries.
2. Keep it as an override of `base_hosts_file` in `containers.conf`.

I'm leaning towards (2) since we have a configuration item in `containers.conf` already. What do you think?

Luap99 commented:
I say we should do 2

rhatdan commented:
@gavinkflam Still working on this?

gavinkflam commented:
@rhatdan @Luap99 This is now complete. Apologies for the delay - it's been an eventful year, and I lost track of this open PR.

Luap99 commented on test/e2e/containers_conf_test.go:
```diff
@@ -463,6 +463,60 @@ var _ = Describe("Verify podman containers.conf usage", func() {
+			})
+
+			It("should not use any hosts files", func() {
+				Expect(session.OutputToString()).ToNot(ContainSubstring("localhost.localdomain"))
+				Expect(session.OutputToString()).To(ContainSubstring("localhost"))
```
Can you do an exact match for
```
127.0.0.1       localhost
::1     localhost
```
This should be more robust I think.

Also one important bit is that we still add our own custom host entries for the host.containers.internal and the containers name/hostname as well. So I think we need to check that as well for full testing (I know it is not really related to your current change and a testing gap from earlier)
If you run the container with --hostname abc or whatever name you like we could always match that this name is in the file for all cases, i.e. still in the JustBeforeEach() function

gavinkflam replied:
Thanks added all these

gavinkflam commented on test/e2e/containers_conf_test.go:
```diff
@@ -463,6 +463,60 @@ var _ = Describe("Verify podman containers.conf usage", func() {
+			})
+
+			It("should use the hosts file from the container image", func() {
+				Expect(session.OutputToString()).To(ContainSubstring("localhost localhost.localdomain"))
```
Good idea. I have added such a custom image to add the related tests.

gavinkflam commented:
> Thanks mostly looks good but I think we can do a bit better on the tests

Thank you for the suggestions, @Luap99! I’ve enhanced the tests and implemented the flag for pod create. The pod create command now uses this flag to configure the infra-container appropriately.

During testing, I discovered that image builds using the `--no-hosts` flag were failing in remote tests. I realized that this flag was never supported in the remote build APIs. To address this, I’ve added a `nohosts` option to the build APIs.

Luap99 commented on libpod/container_config.go:
```diff
@@ -290,15 +290,14 @@ type ContainerNetworkConfig struct {
+	// This must either be an absolute path to a file on the host system, or one of the
+	// special flags `image` or `none`.
+	// If it is empty it defaults to the base_hosts_file configuration in containers.conf.
+	HostsFile string `json:"hostsFile,omitempty"`
```
That is a breaking change, please do not rename that. All containers that had the old key set will all of the sudden be unset which breaks them on restart. It is really no worth to rename that IMO.

gavinkflam replied:
Thanks reverted

Luap99 commented on pkg/specgen/specgen.go:
```diff
@@ -539,18 +539,18 @@ type ContainerNetworkConfig struct {
+	// If it is empty it defaults to the base_hosts_file configuration in containers.conf.
+	// Conflicts with UseImageHosts.
+	// Optional.
+	HostsFile string `json:"hosts_file,omitempty"`
```
Here as well, this is a breaking change as specgen is a stable type exposed in the bindings.

gavinkflam replied:
This is reverted too

gavinkflam commented:
Reverted rename and rebased.

Luap99 reviewed (approved):
LGTM
Thank you

gavinkflam commented:
/packit retest-failed
